### ISSUE TITLE
`cors` route rule should auto response to preflight

### ISSUE THREAD

**Issue by MickL** [nitrojs/nitro#2340]
### Environment

```
nitropack: 2.9.6
```

### Reproduction

https://stackblitz.com/edit/github-xaqmuq?file=nitro.config.ts

### Describe the bug

Setting `cors: true` to does not work:

```ts
routeRules: {
    '/**': {
      cors: true,
    },
},
```

The browser still throws cors error:

> Access to fetch at 'http://localhost:3333/' from origin 'http://localhost:3001' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: No 'Access-Control-Allow-Origin' header is present on the requested resource. If an opaque response serves your needs, set the request's mode to 'no-cors' to fetch the resource with CORS disabled.

With header, too:

```
routeRules: {
    '/**': {
      cors: true,
      headers: {
        'Access-Control-Allow-Origin': '*',
        'Access-Control-Allow-Headers': '*',
      },
    },
  },
```

Error:

> Access to fetch at 'http://localhost:3333/' from origin 'http://localhost:3001' has been blocked by CORS policy: Response to preflight request doesn't pass access control check: It does not have HTTP ok status.

### Additional context

I already opened an [issue at Nuxt](https://github.com/nuxt/nuxt/issues/25686) that setting `corse:true` does not have an effect on `/api` routes, only on `/public` folder.

**Comment by MickL** [comment]
It seems to work when setting the header (as shown above):

```
routeRules: {
    '/**': {
      cors: true,
      headers: {
        'Access-Control-Allow-Origin': '*',
      },
    },
  },
```

But it seems like it does not match too nested routes:

`await $fetch(`/admin/products/${productId}/status`, { method: 'PATCH' });` -> Ok
`await $fetch(`/admin/products/${productId}/images/{$id}`, { method: 'DELETE' });` -> Error
`await $fetch(`/admin/products/${productId}/images/{$id}/order`, { method: 'PATCH' });` -> Error

@pi0 Is this an issue with an UnJS package maybe?

**Comment by MickL** [comment]
Hi @pi0 , I digged deeper into this and found a critical bug with route rules not beeing applied. Maybe a problem with https://github.com/unjs/radix3?

The following rule works for all routes:
```
routeRules: {
    '/**': {
      cors: true,
      headers: {
        'Access-Control-Allow-Origin': '*',
      },
    },
  },
```

BUT if a route does not have GET route, the routeRoule is not applied. E.g.:

```
/products/index.ts
/products/index.patch.ts
/items/index.patch.ts -> Route rule not applied
```

Therefor the routes without a GET route throw a cors error in my case.

I made a reproduction repo, you can simply run `bun install` and then `bun dev` to start a Nuxt app at localhost:3000 and a Nitro app at localhost:3001

https://github.com/MickL/nitro-route-rules-bug

**Comment by MickL** [comment]
It gets even more weird!

The route rules are applied if there is a GET route, e.g. `contact.ts` and `contact.post.ts` BUT if the GET route just throws an error it is NOT applied:

Route rule applied for POST route:
```ts
export default defineEventHandler(async (event) => {
   return '404';
});
```

Route rule NOT applied for POST route:
```ts
export default defineEventHandler(async (event) => {
  throw createError({
    statusCode: 404,
  });
});
```

**Comment by ArcherScript** [comment]
This is insane. I spent the last 8 hours trying to figure out what was going on with my CORS headers not being set. I added a GET route and it fixed all of my problems.

I'm genuinely regretting using nitro.

**Comment by ArcherScript** [comment]
Possibly related to #1806

**Comment by MickL** [comment]
This is definitely a critical bug and an annoying one as well. Maybe it will be fixed with v3? @pi0 

But in any case it needs to be fixed for v2 aswell. This also makes problems within Nuxt: https://github.com/nuxt/nuxt/issues/25686

**Comment by pi0** [comment]
I have created a better sandbox to show Nitro behavior: 

https://stackblitz.com/edit/github-mdtiap?file=server%2Froutes%2Findex.ts,server%2Froutes%2Fapi%2Ftest.post.ts

Nitro actually do apply route rules to all methods regardless if there is a handler registered for them or not. (notice that there are cors headers to the post-only API route's response).


The issue is how CORS preflight works in browsers. Browsers use an `OPTIONS` request to check CORS availability (not even `GET`) and the reason adding `[path].ts` (without method suffix) fixes the problem is that it _also_ handles that. 

I understand it can be frustrating with CORS issues but it isn't a bug at least with how route rules work.

(...so what we can improve...)

I understand DX is not good and CORS issues, god knows how much frustration they cause every developer, and that when Nitro allows to set a magical route rule `cors: true`, we expect that it just works, and it should be.

We can improve our route rule handler to auto-respond to OPTIONS even if there is not a handler but the problem is that it is too soon. Route rules are applied before any of the user handlers and if a user wants to actually handle OPTIONS method of a path (to apply their security policy -- the very purpose of CORS functionality) nitro will implicitly override this. So while it might look like an easy fix/feature, it might cause security implications which are worst than a not working CORS call when it matters.

I will be thinking about how we can safely implement this as a fallback approach (nitro error handler could be one place, by checking cors route rule) but i think one first and good step is that we at least document this:

CORS functionality needs API handlers to respond to OPTIONS requests and as a userland fix, one can add `server/api/[...path].ts` 

```
// server/api/[...path].ts
export default defineEventHandler((event) => {
  if (event.method === 'OPTIONS') {
    return ""
  }
});
```

**Comment by ArcherScript** [comment]
Why do we have 
- https://h3.unjs.io/utils/advanced#appendcorsheadersevent-options
- https://h3.unjs.io/utils/advanced#appendcorspreflightheadersevent-options
- https://h3.unjs.io/utils/advanced#handlecorsevent-options
- https://h3.unjs.io/utils/response#setresponseheadersevent-headers
- https://h3.unjs.io/utils/response#setresponseheaderevent-name-value

Do these only work in a get route file as well?

**Comment by pi0** [comment]
@ArcherScript Preflight request is _not_ GET, it is OPTIONS.

I suggest you read MDN docs about [Preflight request](https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request) and [CORS](https://developer.mozilla.org/en-US/docs/Glossary/CORS) mechanism first.

Adding `server/api/test.ts` adds an event handler that handles _all_ methods (GET, POST, OPTIONS, etc) and fixes the issue regardless of what it responds even an empty string with no headers is enough because route rule already adds `*` cors headers for us.

Browsers expect to have an "ok" response when they want to verify CORS (via OPTIONS fetch). If there is no event handler, Nitro (your application), responds with 404 or 405 both unacceptable by browsers.

**Comment by ArcherScript** [comment]
> @ArcherScript Preflight request is _not_ GET, it is OPTIONS.
> 
> I suggest you read MDN docs about [Preflight request](https://developer.mozilla.org/en-US/docs/Glossary/Preflight_request) and [CORS](https://developer.mozilla.org/en-US/docs/Glossary/CORS) mechanism first.
> 
> Adding `server/api/test.ts` adds an event handler that handles _all_ methods (GET, POST, OPTIONS, etc) and fixes the issue regardless of what it responds even an empty string with no headers is enough because route rule already adds `*` cors headers for us.
> 
> Browsers expect to have an "ok" response when they want to verify CORS (via OPTIONS fetch). If there is no event handler, Nitro (your application), responds with 404 or 405 both unacceptable by browsers.
> 
> You can alternatively use [`handleCors`](https://h3.unjs.io/utils/advanced#handlecorsevent-options) h3 utility (instead of route rule shortcut that simply adds headers) to have better control over CORS strategy but you still need to add it to a catch all handler (without `.post` suffix)

Yeah this is what I thought. Thanks for the explanation.

It seems like `server/api/route.ts` is the ONLY option if you are trying to make a request from the browser. Which is not clear from the docs.

I understand why it is that way though from the explanation.

**Comment by ArcherScript** [comment]
Should we expand the documentation on `handleCors` and other cors helper methods so that it explains how these only function in `route.ts` files?

**Comment by pi0** [comment]
Surely feel free to contribute to h3 docs if you could successfully use it 🙏🏼  [jsdocs is here](https://github.com/unjs/h3/blob/main/src/utils/cors.ts#L89). 

Mind you, nitro route rules cors handler does _not_ use h3 utility btw. it uses simple headers (why? because it can be natively applied to deployment provider route rules). 

I think we might also add a new `## cors` section to the end of [/guide/routing](https://nitro.unjs.io/guide/routing) to mention different ways of handling cors but most importantly, that cors needs a catch all handler.

**Comment by MickL** [comment]
One thing is CORS but this issue actually turned out to a critical bug. Maybe it was not clear enough in the end but it took me some time to figure it out:

Route roules are not applied if a route does not have a GET route. For example:

```
/products.ts
/products.patch.ts
/items.patch.ts -> Route rule not applied
```

Or did I miss something?

**Comment by pi0** [comment]
@MickL please locally run my reproduction from https://github.com/unjs/nitro/issues/2340#issuecomment-2308264327. Route rules **always** apply regardless of incoming request or if there is even a matching router handelr or not. they are applied on their own. CORS fails not because roue rule `cors: true` is not adding headers it fails because browser expects an "ok" (like 200) response and if there is no route handler, well it will be 404/405. A catch all handler makes sure of that 200 (and that i plan to make additional behavior to route rules to be able to auto-response, considering we can be sure of security)

**Comment by MickL** [comment]
But why does it work when adding the get route? Like for PATCH /product it works in my example. Sorry if I miss something.

**Comment by pi0** [comment]
```
/xyz.ts <-- This is a catch-all route (handling ALL methods)
```

- Route rule adds headers for paths it is defined for (and works)
- If there is no route handler that matches the path + method, the server responds 404 (not found) or 405 (method not allowed). 

The browser makes a preflight to the same path with OPTIONS method. It expects 1) headers 2) ok response. route rules does (1), you need to do (2) by adding any kind of catch-all handler that simply gives that "ok".

**Comment by MickL** [comment]
I see, thanks for all the explanations!

**Comment by MickL** [comment]
**Another issue** with this is that non-existent routes return cors error instead of 404.

**Comment by MickL** [comment]
I use a middleware until this is fixed:

**1.cors.ts (Nitro v1)**

```ts
import { defineEventHandler, getMethod, setResponseStatus } from 'h3';

export default defineEventHandler((event) => {
  if (getMethod(event) === 'OPTIONS') {
    setResponseStatus(event, 204);
    return {};
  }
});
```

**1.cors.ts (Nitro v2)**

```ts
import { defineHandler, isMethod } from 'nitro/h3';

export default defineEventHandler((event) => {
  if (isMethod(event, 'OPTIONS')) {
    event.res.status = 204;
    return {};
  }
});
```

**Comment by MinetaS** [comment]
Is there any update on this issue? This issue has been open for over a year but not received much attention as I see.

My case was headers not being set with nuxt-security:
```ts
  routeRules: {
    "/api/management/**": {
      security: {
        corsHandler: {
          origin: "*",
          methods: ["GET", "POST", "OPTIONS"],
          ...
        }
      }
    }
  }
```

Dev server sets the headers properly while production build doesn't.

**Comment by MickL** [comment]
I am not sure if there is anything in the works for Nitro v2 @pi0 , but you can simply add a middleware (no matter if v1 or v2) as I showed [above](https://github.com/nitrojs/nitro/issues/2340#issuecomment-3092276231).

If you want to allow cross origin requests I have an example for a cors middleware here: https://github.com/nitrojs/nitro/issues/2903#issuecomment-3092275129

Otherwise, I dont know what you are trying to do because it seems to be related to nuxt-security, not sure if this is related to his issue here.

**Comment by pi0x** [comment]
Fixed in v3 — the `cors` route rule answers preflights itself (204 with the allow headers) and applies even when the route has no GET handler; it is handled by h3's `handleCors` through h3/rules.

_Triaged with AI assistance._
